frontend/src/main/web/src/components/TextInput.jsx

```diff
@@ -0,0 +1,185 @@
+import React, { Component, PropTypes } from 'react'
+import ReactDOM from 'react-dom'
+import TextareaAutosize from 'react-textarea-autosize'
+import { flattenThemeClasses } from '../utils/styleUtils'
+import { isEqual } from 'lodash'
+
+const classes = {
+  base: {
+    ap: 'Ap(n)',
+    bgc: 'Bgc(t)',
+    bdc: 'Bdc(neutral)',
+    bdw: 'Bdw(2px)',
+    bds: 'Bds(s)',
+    bdrs: 'Bdrs(rq)',
+    bxz: 'Bxz(bb)',
+    c: 'C(i)',
+    ff: 'Ff(inh)',
+    fw: 'Fw(inh)',
+    fz: 'Fz(inh)',
+    fs: 'Fs(inh)',
+    o: 'O(n)',
+    p: 'Px(rq) Py(re)',
+    ph: 'Ph(neutral)',
+    focus: {
+      bdc: 'Bdc(pri):f'
+    }
+  }
+}
+
+class TextInput extends Component {
+  constructor () {
+    super()
+  }
+
+  _onBlur (e) {
+    const { onBlur } = this.props
+    if (onBlur) {
+      onBlur(e)
+    }
+  }
+  _onChange (e) {
+    const { onChange, onChangeText } = this.props
+    if (onChangeText) onChangeText(e.target.value)
+    if (onChange) onChange(e)
+  }
+
+  _onFocus (e) {
+    const { clearTextOnFocus, onFocus, selectTextOnFocus } = this.props
+    const node = ReactDOM.findDOMNode(this)
+    if (clearTextOnFocus) node.value = ''
+    if (selectTextOnFocus) node.select()
+    if (onFocus) onFocus(e)
+  }
+
+  _onSelectionChange (e) {
+    const { onSelectionChange } = this.props
+    const { selectionDirection, selectionEnd, selectionStart } = e.target
+    const event = {
+      selectionDirection,
+      selectionEnd,
+      selectionStart,
+      nativeEvent: e.nativeEvent
+    }
+    if (onSelectionChange) onSelectionChange(event)
```

Comment:
Could move the condition up to the second line so it will only create `event` when it needs to:

```
const { onSelectionChange } = this.props
if (onSelectionChange) {
  // create event
  onSelectionChange(event)
}
```